File: Kalorie/Kalorie/Core/UseCases/ReauthenticateUseCase.swift
//
//  ReauthenticateUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 01.09.2026.
//

import FirebaseAuth
import Foundation

protocol ReauthenticateUseCaseProtocol {
    func callAsFunction() async throws
}

struct ReauthenticateUseCase: ReauthenticateUseCaseProtocol {

    // MARK: - Properties

    private let appleSignInProvider: any AppleSignInProviderProtocol
    private let googleSignInProvider: any GoogleSignInProviderProtocol
    private let authCommandProvider: any AuthCommandProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(
        appleSignInProvider: any AppleSignInProviderProtocol,
        googleSignInProvider: any GoogleSignInProviderProtocol,
        authCommandProvider: any AuthCommandProviderProtocol,
        authProvider: any AuthProviderProtocol
    ) {
        self.appleSignInProvider = appleSignInProvider
        self.googleSignInProvider = googleSignInProvider
        self.authCommandProvider = authCommandProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction() async throws {
        let credential: AuthCredential
        switch authProvider.linkedProviderKind {
        case .apple:
            credential = try await appleSignInProvider.signIn().credential
        case .google:
            credential = try await googleSignInProvider.signIn().credential
        case nil:
            throw AuthError.notAuthenticated
        }
        try await authCommandProvider.reauthenticate(with: credential)
    }
}

#if DEBUG
struct ReauthenticateUseCaseFake: ReauthenticateUseCaseProtocol {

    // MARK: - Properties

    var errorToThrow: Error?

    // MARK: - Functions

    func callAsFunction() async throws {
        if let errorToThrow { throw errorToThrow }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/ReauthenticateUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthCommandProviderProtocol
import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderKind
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.auth.GoogleSignInProviderProtocol

interface ReauthenticateUseCaseProtocol {
    suspend operator fun invoke()
}

class ReauthenticateUseCase(
    private val googleSignInProvider: GoogleSignInProviderProtocol,
    private val authCommandProvider: AuthCommandProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : ReauthenticateUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke() {
        val credential = when (authProvider.linkedProviderKind) {
            AuthProviderKind.GOOGLE -> googleSignInProvider.signIn().credential
            null -> throw AuthError.NotAuthenticated
        }
        authCommandProvider.reauthenticate(credential)
    }
}
